File: programovani_v_jave/cviceni_hnetynka/du/justify/JustifiedWriter.java
package com.mypackage.task4;

import java.io.*;
import java.util.ArrayList;
import java.util.Scanner;

public class JustifiedWriter {

    private PrintStream tw;
    private int lineLength;

    public JustifiedWriter(PrintStream tw, int lineLength)
    {
        this.tw = tw;
        this.lineLength = lineLength;
    }


    /**
     * Writes the text justified to the current line lenght.
     */
    public void WriteJustified(BufferedReader r) throws IOException {
        var words = new ArrayList<String>();
        int wordsCummulativeLength = 0;

        WordReader wr = new WordReader(r);
        Token t = wr.getNextToken();

        while (true)
        {
            if(t  instanceof  EndOfStreamToken) break;
            else if (t instanceof  EndOfParagraphToken)
            {
                WriteLine(words, false);
                EndParagraph();
                words.clear();
                wordsCummulativeLength = 0;
            }
            else if (t instanceof  WordToken) {
                var word = ((WordToken) t).getWord();

                if (wordsCummulativeLength + word.length() > lineLength)
                {
                    if (words.size() != 0)
                    {
                        WriteLine(words, true);
                        words.clear();
                        wordsCummulativeLength = 0;
                    }
                }

                wordsCummulativeLength += word.length() + 1;
                words.add(word);

            }

            t = wr.getNextToken();
        }


        if (words.size() > 0)
            WriteLine(words, false);

    }

    public void WriteLine(ArrayList<String> words, boolean justify) {
        if (justify && words.size() > 1)
        {
            writeLineJustified(words);
        }
        else
            writeLineNotJustified(words);
    }

    private void writeLineJustified(ArrayList<String> words)
    {

        int spaces = this.lineLength - words.stream().mapToInt( word -> word.length()).sum();
        int spaceRegions = (words.size() - 1);
        int spacesPerBlock = spaces / spaceRegions;
        int extraSpaces = spaces % spaceRegions;
        String space = new String(new char[spacesPerBlock]).replace("\0", " ");
        String extraSpace = space + ' ';

        for (var word : words)
        {
            tw.print(word);
            if (spaceRegions > 0)
            {
                spaceRegions--;
                if(extraSpaces > 0)
                {
                    extraSpaces--;
                    tw.print(extraSpace);
                }
                else
                    tw.print(space);
            }
        }

        writeNewLine();

    }

    private void writeLineNotJustified(ArrayList<String> words)
    {
        tw.print(String.join(" ", words));
        writeNewLine();
    }

    private void writeNewLine()  {
        tw.print('\n');
    }
    public void EndParagraph() { writeNewLine(); }



}

File: NPRG035_programovani_v_csharp/03-align-files/JustifiedWriter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace task3;

public class JustifiedWriter
{
    private TextWriter tw;
    private int lineLength;
    private char spaceChar;
    private string? appendage;
    public JustifiedWriter(TextWriter tw, int lineLength, char spaceChar = ' ', string? appendage = null)
	{
        this.tw = tw;
        this.lineLength = lineLength;
        this.spaceChar = spaceChar;
        this.appendage = appendage;
    }


    /**
     * Writes the text justified to the current line lenght.
     */
    public void WriteJustified(TokenReader reader)
    {

        var words = new List<string>();
        int wordsCummulativeLength = 0;

        foreach (Token token in reader.Read())
        {

            if (token is ParagraphBreakToken)
            {
                WriteLine(words, false);
                EndParagraph();
                words.Clear();
                wordsCummulativeLength = 0;
                continue;
            }
            else if (token is WordToken)
            {
                var word = ((WordToken)token).Word;
                if (wordsCummulativeLength + word.Length > lineLength)
                {
                    if (words.Count != 0)
                    {
                        WriteLine(words);
                        words.Clear();
                        wordsCummulativeLength = 0;
                    }
                }

                wordsCummulativeLength += word.Length + 1;
                words.Add(word);
            }
            // ignore file break
        }


        if (words.Count() > 0)
            WriteLine(words, false);

    }

    public void WriteLine(List<string> words, bool justify = true) {
        if (justify && words.Count > 1)
        {
            writeLineJustified(words);
        }
        else
            writeLineNotJustified(words);
    }   

    private void writeLineJustified(List<string> words)
    {
        int spaces = lineLength - words.Select(word => word.Length).Sum();
        int spaceRegions = (words.Count - 1);
        int spacesPerBlock = spaces / spaceRegions;
        int extraSpaces = spaces % spaceRegions;
        string space = string.Concat(Enumerable.Repeat(spaceChar, spacesPerBlock));
        string extraSpace = space + spaceChar;

        foreach (var word in words)
        {
            tw.Write(word);
            if (spaceRegions > 0)
            {
                spaceRegions--;
                if(extraSpaces > 0)
                {
                    extraSpaces--;
                    tw.Write(extraSpace);
                }
                else
                    tw.Write(space);
            }
        }

        writeNewLine();

    }

    private void writeLineNotJustified(List<string> words)
    {
        tw.Write(String.Join(spaceChar, words));
        writeNewLine();
    }

    private void writeNewLine()
    {
        if (appendage is not null) tw.Write(appendage);
        tw.Write('\n');
    }
    public void EndParagraph() => writeNewLine();
}
